managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1144,16 +1144,14 @@ public ManagedCursor newNonDurableCursor(Position startCursorPosition, String cu
             return cachedCursor;
         }
 
-        NonDurableCursorImpl cursor = new NonDurableCursorImpl(bookKeeper, this, cursorName,
-                startCursorPosition, initialPosition, isReadCompacted);
-        cursor.setActive();
-
-        log.info("[{}] Opened new cursor: {}", name, cursor);
         synchronized (this) {
+            NonDurableCursorImpl cursor = new NonDurableCursorImpl(bookKeeper, this, cursorName,
+                    startCursorPosition, initialPosition, isReadCompacted);
+            cursor.setActive();
+            log.info("[{}] Opened new cursor: {}", name, cursor);
             addCursor(cursor);
+            return cursor;
```

Comment:
Since there is no test to protect this change (it might be hard to add the test), to avoid possible regression in futurem you'd better add some comments here to explain why the creation of the `NonDurableCursorImpl` should be put inside the synchronized block.

Reply:
From the PR description, it seems that we should only synchronize the `initializeCursorPosition` method with `internalTrimLedgers`, would it be better to just minimum the synchronize block like:

```java
    void initializeCursorPosition(Pair<Position, Long> lastPositionCounter) {
        synchronized (ledger) {
            /* ... */
        }
    }
```

Reply:
Added